### ISSUE TITLE
Union columns can never be `NULL`

### ISSUE THREAD

**Issue by samuelcolvin** [apache/datafusion#11162]
### Describe the bug

Maybe I'm doing something wrong with my union in [datafusion-functions-json](https://github.com/datafusion-contrib/datafusion-functions-json), but `is null` expressions never evaluate to true with the `JsonUnion` column.

They behave as expected when the function in question returns a scalar `JsonUnion`.

### To Reproduce

See https://github.com/datafusion-contrib/datafusion-functions-json/pull/24

### Expected behavior

When the value in every column of the union is null:
* `<union column> is null` should evaluate to null
* I guess it would also be nice if  the repr of null values in a union was an empty string same as everything else, not `{<first column_name>=}` (which I what I think it is now)

**Comment by samuelcolvin** [comment]
@alamb any idea on where I would start looking to try and fix this?

**Comment by samuelcolvin** [comment]
See #11314 as a demonstration of the problem for both dense and sparse unions.

After a bit of investigation, the issues lies in the first instance with

https://github.com/apache/datafusion/blob/08c5345e932f1c5c948751e0d06b1fd99e174efa/datafusion/physical-expr/src/expressions/is_null.rs#L74-L84

Then with [this code](https://github.com/apache/arrow-rs/blob/b9562b9550b8ff4aa7be9859e56e467b1a3b3de6/arrow-arith/src/boolean.rs#L314-L332) in `arrow-rs`:

```rs
/// Returns a non-null [BooleanArray] with whether each value of the array is null.
/// # Error
/// This function never errors.
/// # Example
/// ...
pub fn is_null(input: &dyn Array) -> Result<BooleanArray, ArrowError> {
    let values = match input.logical_nulls() {
        None => BooleanBuffer::new_unset(input.len()),
        Some(nulls) => !nulls.inner(),
    };

    Ok(BooleanArray::new(values, None))
}
```

And then with [this code](https://github.com/apache/arrow-rs/blob/b9562b9550b8ff4aa7be9859e56e467b1a3b3de6/arrow-array/src/array/union_array.rs#L482-L486)

```rs
    /// Union types always return non null as there is no validity buffer.
    /// To check validity correctly you must check the underlying vector.
    fn is_null(&self, _index: usize) -> bool {
        false
    }
```

Ultimately with [the spec](https://github.com/apache/arrow/blob/674e70891d1b3bc82b025d9c434d8ff1aa4c877e/docs/source/format/Columnar.rst?plain=1#L862-L864)

> Unlike other data types, unions do not have their own validity bitmap. Instead,
> the nullness of each slot is determined exclusively by the child arrays which
> are composed to create the union.

---

Basically arrow is saying "we're not going to tell you if a union is null, you need to look in the child arrays", but datafusion isn't listening and is just asking the union if it's null in the naive way.

Two options to move forward as far as I can tell:
1. Decide unions in DF can never be null — I'll need to abandon unions in `datafusion-functions-json` and just return strings everywhere
2. Have custom logic for unions that looks up the child array to determine if the value is null

If (as I hope) we go for the second option, there's also the issue (as demonstrated by #11314) that the representation of  "null" union items doesn't match other types, it shows `{A=}` instead of an empty string.

**Comment by jayzhan211** [comment]
> Have custom logic for unions that looks up the child array to determine if the value is null
+1 for second option. I think we should check the children's nullability.

**Comment by samuelcolvin** [comment]
I suppose there's a third option of updating arrow-rs to correctly calculate if a `UnionArray` is null, but I presume that works take much longer

**Comment by samuelcolvin** [comment]
I've proposed a fix in #11321.

**Comment by alamb** [comment]
> I suppose there's a third option of updating arrow-rs to correctly calculate if a `UnionArray` is null, but I presume that works take much longer

It would likely take longer

Note there is a method that takes into account child nullability that perhaps we could use instead of `is_null`: 
https://docs.rs/arrow/latest/arrow/array/trait.Array.html#method.logical_nulls
